torch/distributions/von_mises.py

```diff
@@ -0,0 +1,140 @@
+from __future__ import absolute_import, division, print_function
+
+import math
+
+import torch
+import torch.jit
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.utils import broadcast_all, lazy_property
+
+
+def _eval_poly(y, coef):
+    coef = list(coef)
+    result = coef.pop()
+    while coef:
+        result = coef.pop() + y * result
+    return result
+
+
+_I0_COEF_SMALL = [1.0, 3.5156229, 3.0899424, 1.2067492, 0.2659732, 0.360768e-1, 0.45813e-2]
+_I0_COEF_LARGE = [0.39894228, 0.1328592e-1, 0.225319e-2, -0.157565e-2, 0.916281e-2,
+                  -0.2057706e-1, 0.2635537e-1, -0.1647633e-1, 0.392377e-2]
+_I1_COEF_SMALL = [0.5, 0.87890594, 0.51498869, 0.15084934, 0.2658733e-1, 0.301532e-2, 0.32411e-3]
+_I1_COEF_LARGE = [0.39894228, -0.3988024e-1, -0.362018e-2, 0.163801e-2, -0.1031555e-1,
+                  0.2282967e-1, -0.2895312e-1, 0.1787654e-1, -0.420059e-2]
+
+_COEF_SMALL = [_I0_COEF_SMALL, _I1_COEF_SMALL]
+_COEF_LARGE = [_I0_COEF_LARGE, _I1_COEF_LARGE]
+
+
+def _log_modified_bessel_fn(x, order=0):
+    """
+    Returns ``log(I_order(x))`` for ``x > 0``,
+    where `order` is either 0 or 1.
+    """
+    assert order == 0 or order == 1
+
+    # compute small solution
+    y = (x / 3.75)
+    y = y * y
+    small = _eval_poly(y, _COEF_SMALL[order])
+    if order == 1:
+        small = x.abs() * small
+    small = small.log()
+
+    # compute large solution
+    y = 3.75 / x
+    large = x - 0.5 * x.log() + _eval_poly(y, _COEF_LARGE[order]).log()
+
+    result = torch.where(x < 3.75, small, large)
+    return result
+
+
+@torch.jit.script

```

Comment:
I didn't dig too much into the rest of this PR, but would be it OK to remove the `@torch.jit.script` here?

If this function is used in some other TorchScript method/module it will automatically be compiled, and having a `@torch.jit.script` decorator inside the core PyTorch library starts up the jit compiler anytime someone does `import torch` which does some expensive initialization (so it impacts all users of PyTorch instead of only those using the jit)

Reply:
Yes, please go for it. Actually, I didn't realize we aren't supposed to not put `torch.jit.script` decorators in the core code; we should add a lint rule for this.

Reply:
The reason I recommended `@torch.jit.scropt` here was to handle the data-dependent control flow when jit tracing code involving this distribution. We use this pattern of `@torch.jit.script`ing low level bits that cannot be trace, in Pyro. Do you have a recommended workaround in PyTorch core code, such that we can support `torch.jit.trace`?

Reply:
I had a PR that did pretty much what you want (do nothing, but if tracing then compile the function) in #25746, I could revive that as something like `@torch.jit._lazy_script`

Reply:
@driazati sure, a `@torch.jit._lazy_script` would be great.